[GTK][WPE] FPS throttling in DisplayLink

psaavedra opened the pull request:
#### 6230eecf3f6be5d63f1e9265d72f9c7570659613
<pre>
[GTK][WPE] FPS throttling in DisplayLink
<a href="https://bugs.webkit.org/show_bug.cgi?id=301860">https://bugs.webkit.org/show_bug.cgi?id=301860</a>

Reviewed by Nikolas Zimmermann and Carlos Alberto Lopez Perez.

Introduce `WEBKIT_DISPLAY_REFRESH_THROTTLE_FPS` environment variable to
adjust FPS throttling. This variable allows overwrite the refresh rate
used for the observer notifications.

This is particularly useful when running performance benchmarks such as
MotionMark on low-end devices (for example, Raspberry Pi 4). On such devices,
the default frame rate often leads to extremely low scores (1-3 pts), making
it difficult to track performance regressions. By throttling the target
frame rate to a lower value (e.g. 15 FPS), the benchmark produces results in
a more useful range (around 100-600 pts). This improves the comparability and
regression analysis. It also improves the quality of animations by making
their complexity more visible, even if they run at a lower frame rate.

* Source/WebKit/UIProcess/DisplayLink.h:
* Source/WebKit/UIProcess/glib/DisplayLinkGLib.cpp:
(WebKit::DisplayLink::platformInitialize):

Canonical link: <a href="https://commits.webkit.org/303622@main">https://commits.webkit.org/303622@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/279a90cd517045df6173e7d7a2bae475238abdcb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133017 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5518 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44115 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140553 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85049 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6020 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5381 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101721 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69038 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135963 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119192 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82520 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/4118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1682 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83788 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/113181 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143206 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5187 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37890 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110099 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5269 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4444 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110281 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27958 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3982 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115453 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58772 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5241 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33803 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5082 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68693 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5331 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5199 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann reviewed (approved):
LGTM! @carlosgcampos - can you please check as well?

carlosgcampos commented on Source/WebKit/UIProcess/glib/DisplayLinkGLib.cpp:
```diff
@@ -25,20 +25,34 @@
 
 void DisplayLink::platformInitialize()
 {
+    if (const char* envString = getenv("WEBKIT_PROCESS_DISPLAY_LINK_THROTTLE")) {
```
I would use a different name, I don't understand neither what this means nor what I'm supposed to pass just by looking at the name. It seems this receives a ratio that we use to calculate the refresh rate, right? Wouldn't it be easier for the user to pass the desired fps directly and then calculate the ratio? I guess it's because we don't really support any fps, but only a fraction of the actual one, right? In that case I would still use a better name also avoiding display link which is an internal name that it's not very clear in non apple ports. What about WEBKIT_DISPLAY_REFRESH_THROTTLE_RATIO?

carlosgcampos commented on Source/WebKit/UIProcess/glib/DisplayLinkGLib.cpp:
```diff
@@ -25,20 +25,34 @@
 {
+    if (const char* envString = getenv("WEBKIT_PROCESS_DISPLAY_LINK_THROTTLE")) {
+        if (auto newValue = parseInteger<unsigned>(StringView::fromLatin1(envString)))
+            m_fpsThrottleRatio = *newValue;
```
Should we do any validation of the input? do we allow any value? what happens if 0 is passed, for example?

clopez commented on Source/WebKit/UIProcess/glib/DisplayLinkGLib.cpp:
```diff
@@ -25,20 +25,34 @@
     });
 
-    m_displayNominalFramesPerSecond = m_vblankMonitor->refreshRate();
+    m_displayNominalFramesPerSecond = m_vblankMonitor->refreshRate() / m_fpsThrottleRatio;
```
Wouldn't it better to pass in the environment variable a final value (e.g, 15FPS) rather than a ratio?

By passing a ratio the final value you get is not completely defined as it depends on the value that the GPU and the display have negotiated (can be 30FPS, 60FPS, 120FPS or more)

carlosgcampos replied:
Because we still want to sync with the hardware vblank. If we use a ratio we know we just have to skip frames, but to use a custom fps we would need to fallback to use a timer and lose the hardware synchronization.

nikolaszimmermann replied:
Agreed.

psaavedra replied:
I found an alternative way to make this more aligned with @clopez idea. Now the envar expects FPS(`WEBKIT_PROCESS_DISPLAY_LINK_THROTTLE_FPS`) but only accepts those that are real factor of the Display real refresh rate. That is for a 60Hz display -> 1, 2, 3, 4, 5, 6, ... , 20, 30, 60. This is because the vblank notifications must meet a constant pace.

aperezdc commented on Source/WebKit/UIProcess/glib/DisplayLinkGLib.cpp:
```diff
@@ -35,10 +37,28 @@ void DisplayLink::platformInitialize()
+        if (auto newValue = parseInteger<unsigned>(StringView::fromLatin1(envString))) {
+            unsigned refreshRate = m_vblankMonitor->refreshRate();
+            // The throttle FPS must be a factor of the real display refresh rate
+            if (!(refreshRate % *newValue) && 0 < *newValue < refreshRate)
```
The `lowerBound < value < upperBound` construct is a Pythonishm 🐍. It should be something like the following:


```suggestion
            if (!(refreshRate % *newValue) && *newValue > 0 && *newValue < refreshRate)
```

carlosgcampos commented on Source/WebKit/UIProcess/glib/DisplayLinkGLib.cpp:
```diff
@@ -35,10 +37,28 @@ void DisplayLink::platformInitialize()
     });
 
     m_displayNominalFramesPerSecond = m_vblankMonitor->refreshRate();
+    if (const char* envString = getenv("WEBKIT_PROCESS_DISPLAY_LINK_THROTTLE_FPS")) {
```
My previous comment about the name still applies. I don't know process means here and I would avoid using display link.

nikolaszimmermann commented on Source/WebKit/UIProcess/glib/DisplayLinkGLib.cpp:
```diff
@@ -35,10 +37,30 @@ void DisplayLink::platformInitialize()
     m_displayNominalFramesPerSecond = m_vblankMonitor->refreshRate();
+    if (const char* envString = getenv("WEBKIT_DISPLAY_REFRESH_THROTTLE_FPS")) {
+        if (auto newValue = parseInteger<unsigned>(StringView::fromLatin1(envString))) {
+            unsigned refreshRate = m_vblankMonitor->refreshRate();
```
I would move this variable to before the m_displayNominalFramesPerSecond expression mark it const and use it everywhere below.

nikolaszimmermann commented on Source/WebKit/UIProcess/glib/DisplayLinkGLib.cpp:
```diff
@@ -35,10 +37,30 @@ void DisplayLink::platformInitialize()
+                WTFLogAlways("WEBKIT_DISPLAY_REFRESH_THROTTLE_FPS=%s rejected: not a factor of refresh rate %ufps (remainder %u)", envString, refreshRate, refreshRate % *newValue); // NOLINT
+        } else
+            WTFLogAlways("WEBKIT_DISPLAY_REFRESH_THROTTLE_FPS=%s rejected: not an integer", envString); // NOLINT
+
```
Trailing newline

nikolaszimmermann reviewed (commented):
Looks good just a small comment

nikolaszimmermann reviewed (approved):
Looks great. Thanks Pablo.

psaavedra committed:
[GTK][WPE] FPS throttling in DisplayLink
https://bugs.webkit.org/show_bug.cgi?id=301860

Reviewed by Nikolas Zimmermann and Carlos Alberto Lopez Perez.

Introduce `WEBKIT_DISPLAY_REFRESH_THROTTLE_FPS` environment variable to
adjust FPS throttling. This variable allows overwrite the refresh rate
used for the observer notifications.

This is particularly useful when running performance benchmarks such as
MotionMark on low-end devices (for example, Raspberry Pi 4). On such devices,
the default frame rate often leads to extremely low scores (1-3 pts), making
it difficult to track performance regressions. By throttling the target
frame rate to a lower value (e.g. 15 FPS), the benchmark produces results in
a more useful range (around 100-600 pts). This improves the comparability and
regression analysis. It also improves the quality of animations by making
their complexity more visible, even if they run at a lower frame rate.

* Source/WebKit/UIProcess/DisplayLink.h:
* Source/WebKit/UIProcess/glib/DisplayLinkGLib.cpp:
(WebKit::DisplayLink::platformInitialize):

Canonical link: https://commits.webkit.org/303622@main
